File: features/dd-sdk-android-trace/src/main/java/com/datadog/legacy/trace/common/sampling/DeterministicSampler.java
/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2016-Present Datadog, Inc.
 */

package com.datadog.legacy.trace.common.sampling;

import com.datadog.opentracing.DDSpan;
import com.datadog.opentracing.DDTracer;

import java.math.BigDecimal;
import java.math.BigInteger;

/**
 * This implements the deterministic sampling algorithm used by the Datadog Agent as well as the
 * tracers for other languages
 */
public class DeterministicSampler implements RateSampler {
    private static final BigInteger KNUTH_FACTOR = new BigInteger("1111111111111111111");
    private static final BigDecimal SPAN_ID_MAX_AS_BIG_DECIMAL =
            new BigDecimal(DDTracer.TRACE_ID_64_BITS_MAX);
    private static final BigInteger MODULUS = new BigInteger("2").pow(64);

    private final BigInteger cutoff;
    private final double rate;

    public DeterministicSampler(final double rate) {
        this.rate = rate;
        cutoff = new BigDecimal(rate).multiply(SPAN_ID_MAX_AS_BIG_DECIMAL).toBigInteger();
    }

    @Override
    public boolean sample(final DDSpan span) {
        final boolean sampled;
        if (rate == 1) {
            sampled = true;
        } else if (rate == 0) {
            sampled = false;
        } else {
            // we are going to use the spanId here instead as it is 64 bits and will not affect the
            // sampling decision. This is the same approach used by the new CoreTracer code.
            sampled = span.getSpanId().multiply(KNUTH_FACTOR).mod(MODULUS).compareTo(cutoff) < 0;
        }

        return sampled;
    }

    @Override
    public double getSampleRate() {
        return rate;
    }
}

File: dd-sdk-android-core/src/main/kotlin/com/datadog/android/core/sampling/DeterministicSampler.kt
/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2016-Present Datadog, Inc.
 */

package com.datadog.android.core.sampling

import androidx.annotation.FloatRange
import com.datadog.android.api.InternalLogger

/**
 * [Sampler] with the given sample rate using a deterministic algorithm for a stable
 * sampling decision across sources.
 *
 * @param T the type of items to sample.
 * @param idConverter a lambda converting the input item into a stable numerical identifier
 * @param sampleRateProvider Provider for the sample rate value which will be called each time
 * the sampling decision needs to be made. All the values should be on the scale [0;100].
 */
open class DeterministicSampler<T : Any>(
    private val idConverter: (T) -> ULong,
    private val sampleRateProvider: () -> Float
) : Sampler<T> {

    /**
     * Creates a new instance lof [DeterministicSampler] with the given sample rate.
     *
     * @param idConverter a lambda converting the input item into a stable numerical identifier
     * @param sampleRate Sample rate to use.
     */
    constructor(
        idConverter: (T) -> ULong,
        @FloatRange(from = 0.0, to = 100.0) sampleRate: Float
    ) : this(idConverter, { sampleRate })

    /**
     * Creates a new instance of [DeterministicSampler] with the given sample rate.
     *
     * @param idConverter a lambda converting the input item into a stable numerical identifier
     * @param sampleRate Sample rate to use.
     */
    constructor(
        idConverter: (T) -> ULong,
        @FloatRange(from = 0.0, to = 100.0) sampleRate: Double
    ) : this(idConverter, sampleRate.toFloat())

    /** @inheritDoc */
    override fun sample(item: T): Boolean {
        val sampleRate = getSampleRate()

        return when {
            sampleRate >= SAMPLE_ALL_RATE -> true
            sampleRate <= 0f -> false
            else -> {
                val hash = idConverter(item) * SAMPLER_HASHER
                val threshold = (MAX_ID.toDouble() * sampleRate / SAMPLE_ALL_RATE).toULong()
                hash < threshold
            }
        }
    }

    /** @inheritDoc */
    override fun getSampleRate(): Float {
        val rawSampleRate = sampleRateProvider()
        return if (rawSampleRate < 0f) {
            InternalLogger.UNBOUND.log(
                InternalLogger.Level.WARN,
                InternalLogger.Target.USER,
                { "Sample rate value provided $rawSampleRate is below 0, setting it to 0." }
            )
            0f
        } else if (rawSampleRate > SAMPLE_ALL_RATE) {
            InternalLogger.UNBOUND.log(
                InternalLogger.Level.WARN,
                InternalLogger.Target.USER,
                { "Sample rate value provided $rawSampleRate is above 100, setting it to 100." }
            )
            SAMPLE_ALL_RATE
        } else {
            rawSampleRate
        }
    }

    private companion object {
        const val SAMPLE_ALL_RATE = 100f

        // Good number for Knuth hashing (large, prime, fit in 64 bit long)
        private const val SAMPLER_HASHER: ULong = 1111111111111111111u

        private const val MAX_ID: ULong = 0xFFFFFFFFFFFFFFFFUL
    }
}
